enhancements/kubernetes-nmstate-integration.md

```diff
@@ -0,0 +1,124 @@
+
+## Summary
+
+A proposal to deploy Kubernetes nmstate on OpenShift.
+
+Node-networking configuration driven by Kubernetes and executed by
+[nmstate](https://nmstate.github.io/).
+
+## Motivation
+
+With hybrid clouds, node-networking setup is becoming even more challenging.
+Different payloads have different networking requirements, and not everything
+can be satisfied as overlays on top of the main interface of the node (e.g.
+SR-IOV, L2, other L2).
+The [Container Network Interface](https://github.com/containernetworking/cni)
+(CNI) standard enables different
+solutions for connecting networks on the node with pods. Some of them are
+[part of the standard](https://github.com/containernetworking/plugins), and there are
+others that extend support for [Open vSwitch bridges](https://github.com/kubevirt/ovs-cni),
+[SR-IOV](https://github.com/hustcat/sriov-cni), and more...
+
+However, in all of these cases, the node must have the networks setup before the
+pod is scheduled. Setting up the networks in a dynamic and heterogenous cluster,
+with dynamic networking requirements, is a challenge by itself - and this is
+what this project is addressing.
+
+### Goals
+
+- Deploy Kubnetes-nmstate as part of openshift
+
+### Non-Goals
+
+- Replace SRIOV operator
+
+## Proposal
+
+A new Kubernetes-nmstate handler DaemonSet is deployed in the cluster part of the OpenShift installation.
+This DaemonSet contains nmstate package and interacts with the NetworkManager
+on the host by mounting the related dbus. The project contains two
+Custom Resource Definitions, `NodeNetworkState` and `NodeNetworkConfigurationPolicy`.
+`NodeNetworkState` objects are created per each node in the cluster and can be
+used to report available interfaces and network configuration. These objects
+are created by kubernetes-nmstate and must not be touched by a user.
+`NodeNetworkConfigurationPolicy` objects can be used to specify desired
+networking state per node or set of nodes. It uses API similar to `NodeNetworkState`.
+
+Kubernetes-nmstate DaemonSet creates a custom resource of `NodeNetworkState` type per each node and
+updates the network topology from each OpenShift node.
+
+User configures host network changes and apply a policy in `NodeNetworkConfigurationPolicy` custom
+resource. Network topology is configured via `desiredState` section in `NodeNetworkConfigurationPolicy`.
+Multiple `NodeNetworkConfigurationPolicy` custom resources can be created, Kubenetes-nmstate DaemonSet will merge
+all the `NodeNetworkConfigurationPolicy` custom resources according to priority definition and
+apply one profile that best matches with the node and update the `desiredState` section in `NodeNetworkState`.
+
+Upon receiving a notification event of `NodeNetworkState` update,
+Kubenetes-nmstate Daemon verify the correctness of `NodeNetworkState` custom resource and
+apply the selected profile to the specific node.
+
+### User Stories
```

Comment:
There is a use case where user doesn't want to assign SR-IOV VFs to a SR-IOV Pod, instead they'd like to use macvlan on top of VFs. Do you think kubernetes-nmstate can be used in such case to manage this kind of VFs once they are created by SR-IOV Operator? By managing VFs, I mean configuring network attrributes such as vlan, mtu etc on that VF device (which, in this case, can be considered as a host device) .

Reply:
AFAIK nmstate does not support configuring these attributes of VF. @EdDev?

Reply:
A VF will eventually become a regular interface, so you could define whatever you want on it.
In addition, https://github.com/nmstate/nmstate/pull/648 extends PF and VF definition to expose other capabilities.

If the PF or VF is defined by a different tool (CNI?), nmstate will "see" it but will consider it "down". You can take control over them using nmstate and define whatever you want.

Reply:
Thanks for the linked PR for supporting PF/VF capabilites via nmstate.
Current SR-IOV Operator is responsible for VF provisioning (creating number of VFs on the host) and manages SR-IOV sub-components, such as SR-IOV CNI and SR-IOV Device Plugin. but we are adding VF index support in SR-IOV Operator to allow it only manages sub-set of VFs device from PF. This, once merged, will leave the rest of VFs on the same PF become un-managed where I see kubernetes-nmstate can fit and take over the management of the rest VFs.  

SR-IOV CNI managed via SR-IOV Operator contains the capability to configure VF properties such as spoof check, trusted VF, MAC address, link state and MTU etc which are the same as Kubernetes-nmstate. But it will only be used when VF is requested and attached to a Pod.  I think this is a clear dividing line that we shall keep in mind going forward between use of kubernetes-nmstate and SR-IOV Operator for configuring VF properties. If it's for Pod VF configuration, then SR-IOV CNI shall be used. If it's for host-level VF configuration that may be used for other purpose instead of directly used in SR-IOV Pod, then Kubernetes-nmstate shall be used.

Several questions regarding provisioning VFs (setting number of VFs ):
1) Is Kubernetes-nmstate also responsible for driver binding such as vfio-pci? or will it be ?
2) When provision VFs for Mellanox card, does it support confguring NIC firmware to set the max number of total_numvfs? Further to this,  do you see Kubernetes-nmstate can be extended to do vendor specific config, such as changing link type (ethernet or infiniband) for mellanox card?

Reply:
> 1. Is Kubernetes-nmstate also responsible for driver binding such as vfio-pci? or will it be ?

DPDK and 3rd party interfaces have been discussed as part of OpenStack requirements.
The discussions and requirements have not materialized into any plan yet, mainly because there was not enough demand for it.
If you are interested in exposing and managing this through nmstate, I suggest pushing it through the regular channels to raise its priority.
I hope we will implement a pluggable provider interface in nmstate in the near future (https://nmstate.atlassian.net/browse/NMSTATE-262), then one can develop and expose custom data using plugins (which can later be embedded into the formal support of nmstate).

> 2. When provision VFs for Mellanox card, does it support confguring NIC firmware to set the max number of total_numvfs?

Current implementation uses NetworkManager as the provider to change these. If it supports it, nmstate and knmstate supports it.
If it does not support it, a BZ needs to be opened in front of NM.
(in special cases, nmstate can extend missing parts in NM as well. We do so already.)

> Further to this, do you see Kubernetes-nmstate can be extended to do vendor specific config, such as changing link type (ethernet or infiniband) for mellanox card?
> 

Seems to fit the discussions we had with OpenStack on 3rd party interfaces.
If the vendor HW fits an existing type, like `ethernet`, then nmstate can expose an additional subtree for specific capabilities of the interface.
As with my DPDK answer, I hope the plugin option will allow vendors specific config could be exposed easily by writing a plugin and later to consider it for full nmstate support.

Reply:
> > 1. Is Kubernetes-nmstate also responsible for driver binding such as vfio-pci? or will it be ?
> 
> DPDK and 3rd party interfaces have been discussed as part of OpenStack requirements.
> The discussions and requirements have not materialized into any plan yet, mainly because there was not enough demand for it.
> If you are interested in exposing and managing this through nmstate, I suggest pushing it through the regular channels to raise its priority.
> I hope we will implement a pluggable provider interface in nmstate in the near future (https://nmstate.atlassian.net/browse/NMSTATE-262), then one can develop and expose custom data using plugins (which can later be embedded into the formal support of nmstate).

Thanks for the reply!
I'm trying to understand, both short term(4.4) and long term(4.5+), the possible duplicate functions between SR-IOV Operator and NMstate Operator and which one shall be used to provision VFs in case of both Operators are installed in the same cluster. Binding of vfio-pci driver to VF interface is needed to allow use of Intel DPDK inside container. If it is not planned for NMstate to support this binding, we shall suggest user to use SR-IOV Operator for such case and same for Mellanox case, if NMstate is not planned to support configuring firmware(vendor specific) on Mellanox card, then it cannot be used for provisioning VFs on Mellanox cards.

Btw, What is the use case that user would use NMstate to provision VFs?

> 
> > 1. When provision VFs for Mellanox card, does it support confguring NIC firmware to set the max number of total_numvfs?
> 
> Current implementation uses NetworkManager as the provider to change these. If it supports it, nmstate and knmstate supports it.
> If it does not support it, a BZ needs to be opened in front of NM.
> (in special cases, nmstate can extend missing parts in NM as well. We do so already.)
> 
Ack.
> > Further to this, do you see Kubernetes-nmstate can be extended to do vendor specific config, such as changing link type (ethernet or infiniband) for mellanox card?
> 
> Seems to fit the discussions we had with OpenStack on 3rd party interfaces.
> If the vendor HW fits an existing type, like `ethernet`, then nmstate can expose an additional subtree for specific capabilities of the interface.
> As with my DPDK answer, I hope the plugin option will allow vendors specific config could be exposed easily by writing a plugin and later to consider it for full nmstate support.

Ack, I asked this in favor of another ongoing discussion about supporting management of RDMA devices in openshift cluster. Since it's hardware related (currently Mellanox), it will require vendor specific settings such as changing device link type from ethernet to infiniband. It looks like nmstate could be a candidate for such settings given the plugin option.

Reply:
> If it is not planned for NMstate to support this binding, we shall suggest user to use SR-IOV Operator for such case and same for Mellanox case, if NMstate is not planned to support configuring firmware(vendor specific) on Mellanox card, then it cannot be used for provisioning VFs on Mellanox cards.

Nmstate was born to offload the host network configuration from layered products such that it will be implemented once and all can share it (instead of each implementing and maintaining it independently).
Therefore, if these tasks are implemented in nmstate, it has the potential to be reused by other projects (OpenStack, oVirt are some examples).
As nmstate is a library, you can use it from whatever operator that fits your need.
IMO, implementing the "how" again in another code base is wasteful, but I may not see the whole picture.
If the declerative nature of the nmstate api fits your needs, I would recommend investing the effort there and gain the shared support advantage.

> Btw, What is the use case that user would use NMstate to provision VFs?

OpenStack and oVirt are pushing VF/s into VM/s, which is common with VM/s in general (not only for those products). The host networking is defining the resources on the host and the virt network part is consuming them. As the resources are host based, they are managed by the host services and the virt part only consumes them (but does not control their life cycle).
For example, one can define a RED network on a host and assign it to a pool of VF/s that connect to the same network. Then, a VM comes up and requests to connect to the RED network and the hypervisor will provide a VF from the relevant pool.

Applications may also take advantage of VF/s for data path acceleration, but these scenarios are less common.

Reply:
> As nmstate is a library, you can use it from whatever operator that fits your need.
> IMO, implementing the "how" again in another code base is wasteful, but I may not see the whole picture.

Thanks for the advice! The fact I'm having this question is SR-IOV Operator has implemented VF provisioning functions & GA the support in 4.3 and I didn't see an equivalent in nmstate at this point. But I'm seeing nmstate as a  project to converge the implementation of VF configuration/provosioning in future releases on OpenShift. Until we get to it, there might be issues that we need to pay attention to for co-existance of VF provision functions in both nmstate and sriov operator.  One example I can think of is that if user configures number of VFs(with different number) from both Operators which all work in declarative ways, then node enters a infinite loop of being cordoned (SR-IOV Operator cordons the node before re-provisioning VFs).

> If the declerative nature of the nmstate api fits your needs, I would recommend investing the effort there and gain the shared support advantage.
> 
> > Btw, What is the use case that user would use NMstate to provision VFs?
> 
> OpenStack and oVirt are pushing VF/s into VM/s, which is common with VM/s in general (not only for those products). The host networking is defining the resources on the host and the virt network part is consuming them. As the resources are host based, they are managed by the host services and the virt part only consumes them (but does not control their life cycle).
> For example, one can define a RED network on a host and assign it to a pool of VF/s that connect to the same network. Then, a VM comes up and requests to connect to the RED network and the hypervisor will provide a VF from the relevant pool.

Ack, I think we use different way to define and manage the VF resource pools on Kubernetes. But that doesn't affect us using the same nmstate library for host-level VF provisioning and configration.

> 
> Applications may also take advantage of VF/s for data path acceleration, but these scenarios are less common.

Can we list this as one user story in favor of supporting V*** case?
For example, `Manage/Configure host SR-IOV network device`, this may include configuration of vlan, mtu on VF device on the host.

Reply:
Done @zshi-redhat can you please take another look?

Reply:
Hi @SchSeba, the change looks good to me.
Nit: manage -> managed.  vlans, mtu drive <- do you mean driver here?

Reply:
you are right my bad thanks!

Reply:
@SchSeba @zshi-redhat can you modify the PR to include the VF-related use case? The discussion above is quite long and involves plenty of implementation and calendar concerns, so I ended up not understanding the motivation. Do you want to create the VF? Only set its vlan/mac/mtu/ip? Why do you want to do that rather than use the PF?